Add and document example data pipeline (#1)
UCLH-Foundry/FlowEHR-Data-Pot: README.md

# Example Data Pipeline on FlowEHR

This repository provides an example of how to create a Data Pipeline that can be deployed as part of [FlowEHR](https://github.com/UCLH-Foundry/FlowEHR).

Using this example, you can build and deploy data transformation code and test it locally via unit tests. Then, you can test it within your dev environment in Azure. Finally, it can be integrated with your deployment of FlowEHR.

## Prerequisites

The following guide assumes you know how to create a Data Pipeline on Databricks and PySpark. It also assumes general familiarity with Azure, Devcontainers, Makefile, Azure Data Factory, and Terraform.

## Local development

For local development, there is a local installation of Spark provided in the Devcontainer image of this repository.  
To start using it, [start your Devcontainer environment](https://code.visualstudio.com/docs/devcontainers/tutorial).

Now you can run unit tests for the Data Pipeline. In the Terminal, run

```Makefile
make test
```

The source code for tests can be found in [./hello-world/tests/](./hello-world/tests/).

> It is highly recommended to write unit tests for all transforms of the data pipeline individually.

Reviewer: ```suggestion
You are strongly urged to write individual unit tests for each transform in the data pipeline.
```

`dbfs:/pipelines/{pipeline_name}/artifacts/{artifact_name}`. 

The rest can be copied over from the example above.

2. A `Makefile` at the root of the repository with an `artifacts` target that will build any artifacts that the pipeline needs to run and put all files that need uploading into `artifacts` directory at the root of the pipeline.

See [example](./hello-world/Makefile) in the repo:

```Makefile
artifacts: build
	mkdir -p artifacts
	cp entrypoint.py artifacts
	cp dist/*.whl artifacts/
```

For a PySpark pipeline, it is recommended to put any library or shared code into a .whl file, as this way it is more straightforward to write automated tests for. There is an [entrypoint.py](./hello-world/entrypoint.py) that is required by Azure Data Factory to be present that is just executing the code that is present in the library.

Note that the name of the artifact must match exactly the name specified in `activities.json`.

## Deploy the pipeline code into a dev environment 

To deploy the pipeline in the context of FlowEHR, please follow these steps:

1. Clone [FlowEHR repository](https://github.com/UCLH-Foundry/FlowEHR).

> 1. Follow the [steps outlined in the README](https://github.com/UCLH-Foundry/FlowEHR#getting-started) to set up your dev environment and deploy FlowEHR. As part of this, you will need to run `make deploy`. 

Reviewer 2: Is `make deploy` now `make infrastructure-transform`?

1. Use one of the following options to reference your pipeline code in FlowEHR repo:

    * Copy over your pipeline code (together with a Makefile and an `artifact.json`) into `/transform/pipelines` into your checked-out copy of FlowEHR. (Note that `/transform/pipelines` is gitignored in FlowEHR). You should have a directory tree similar to the following:

    ```
    └── transform
        └── pipelines
            ├── my-pipeline 
            │   ├── Makefile
            │   ├── activities.json
            │   ├── src
            │   │   └── ...
            │   ├── tests
            │   │   └── ...
            │   └── ...
    ```

    * In `config.transform.yaml` at the root of the FlowEHR repository, provide a link to your Data Pipeline repository. Then, rebuild your Devcontainer (repositories are checked out on Devcontainer startup).
    ```
    ---
    spark_version: 3.3.1
    repositories:
     - git@github.com:tanya-borisova/my-pipeline.git
    ```

> 1. Re-run `make deploy` from the root of FlowEHR while running in your Devcontainer. In the log output, you should see new resources of type `azurerm_data_factory_pipeline.pipeline` being created. 

Reviewer 2: As above, I think this is `make infrastructure-transform` now.
